### PR TITLE
fix: Track graphs of item dependencies to find dependency cycles

### LINKED ISSUE #1122: Build dependency graph between variables in name resolution
### Aim

Trying to write a program
```rs
struct ShouldError {
    uses_nested_arrays: [Array; 2],
}

struct Array {
    array: [Field; 3],
}
```

### Expected behavior

Noir should issue an error that nested arrays are currently unimplemented

### Bug

The error is only issued sometimes when the program is compiled. This is because on some runs `Array` is resolved first and when `ShouldError` is subsequently resolved, it can see all the fields of `Array` and will error that nested arrays are used. On other runs however, `ShouldError` is resolved first and sees only an empty struct for `Array` and thus does not issue an error. In this case, the program continues until monomorphization where it fails with an unreachable assert.

### To reproduce

1.
2.
3.
4.

### Installation method

None

### Nargo version

_No response_

### @noir-lang/noir_wasm version

_No response_

### @noir-lang/barretenberg version

_No response_

### @noir-lang/aztec_backend version

_No response_

### Additional context

_No response_

### Submission Checklist

- [ ] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #1122

## Summary\*

Adds a dependency graph to catch cases where structs will cyclically depend on one other.

For example, the program:

```rs
struct Foo {
    bar: Bar,
}

struct Bar {
    foo: Foo,
}
```
Will now give the following error:
```
error: Dependency cycle found
  ┌─ /home/user/Code/Noir/noir/short/src/main.nr:6:1
  │  
6 │ ╭ struct Bar {
7 │ │     foo: Foo,
8 │ │ }
  │ ╰─' 'Bar' recursively depends on itself: Bar -> Foo -> Bar
  │  
```
The error is still issued if these are split across separate modules.

## Additional Context

Note that self-referential structs like `struct Foo { foo: Foo }` are not currently caught - we may need a separate check for this.

This error is also only currently possible for cyclic struct dependencies, but once this is merged I'm going to expand this to more PRs to allow for globals and type aliases as well. This would implement https://github.com/noir-lang/noir/issues/1734 as well as a similar feature to allow type aliases to reference other type aliases which is currently disallowed.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
Can you open up an issue for self referential structs?

File: compiler/noirc_frontend/src/node_interner.rs [hunk 1]
```diff
@@ -51,6 +58,15 @@ pub struct NodeInterner {
     // Contains the source module each function was defined in
     function_modules: HashMap<FuncId, ModuleId>,
 
+    /// This graph tracks dependencies between different global definitions.
+    /// This is used to ensure the absense of dependency cycles for globals
```
Comment (kevaundray) [inline compiler/noirc_frontend/src/node_interner.rs:62]:
```suggestion
    /// This is used to ensure the absence of dependency cycles for globals
```

**Comment by jfecher** [comment]
@kevaundray there is already https://github.com/noir-lang/noir/issues/3490

**Review by kevaundray — APPROVED** [review]
Generally looks okay

**Pull request merged**